src/clr-angular/data/datagrid/datagrid.ts

```diff
@@ -241,59 +236,85 @@ export class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, On
-          this.renderer.removeClass(this.el.nativeElement, 'datagrid-calculate-mode');
-          this.columns.forEach(column => {
-            this._projectedDisplayColumns.insert(column._view);
-          });
-          this.rows.forEach(row => {
-            this._displayedRows.insert(row._view);
-          });
-        } else {
-          // Set state, style for the datagrid to CALCULATE and insert row & columns into containers
-          this.renderer.addClass(this.el.nativeElement, 'datagrid-calculate-mode');
-          this.columns.forEach(column => {
-            this._projectedCalculationColumns.insert(column._view);
-          });
-          this.rows.forEach(row => {
-            this._calculationRows.insert(row._view);
-          });
-        }
-      })
+      this.emitOnStateProviderChange(),
+      this.emitOnSelectionChange(),
+      this.focusDgTableOnPageChange(),
+      this.resetViewsOnDisplayModeChange(),
+      this.reorderOnRequest()
     );
   }
 
+  private focusDgTableOnPageChange() {
+    return this.page.change.subscribe(() => {
+      this.datagridTable.nativeElement.focus();
+    });
+  }
+
+  private emitOnStateProviderChange(): Subscription {
+    return this.stateProvider.change.subscribe(state => this.refresh.emit(state));
+  }
+
+  private emitOnSelectionChange(): Subscription {
+    return this.selection.change.subscribe(s => {
+      if (this.selection.selectionType === SelectionType.Single) {
+        this.singleSelectedChanged.emit(<T>s);
+      } else if (this.selection.selectionType === SelectionType.Multi) {
+        this.selectedChanged.emit(<T[]>s);
+      }
+    });
+  }
+
+  private displayRowViewsOnRowsChange(): Subscription {
+    return this.rows.changes.subscribe(() => {
+      if (!this.items.smart) {
+        this.items.all = this.rows.map((row: ClrDatagridRow<T>) => row.item);
+      }
+      ViewManagerUtils.detachAllViews(this._displayedRows);
+      ViewManagerUtils.insertAllViews(this._displayedRows, this.rows.toArray());
+    });
+  }
+
+  private resetViewsOnDisplayModeChange(): Subscription {
+    const viewContainers = [
+      this._projectedDisplayColumns,
+      this._projectedCalculationColumns,
+      this._calculationRows,
+      this._displayedRows,
+    ];
+    return this.displayMode.view.subscribe(viewChange => {
+      viewContainers.forEach(viewContainer => ViewManagerUtils.detachAllViews(viewContainer));
+      if (viewChange === DatagridDisplayMode.DISPLAY) {
+        this.renderer.removeClass(this.el.nativeElement, CALCULATE_MODE_CLASS);
+        this.displayViewsInMode(this._projectedDisplayColumns, this._displayedRows);
+      } else {
+        this.renderer.addClass(this.el.nativeElement, CALCULATE_MODE_CLASS);
+        this.displayViewsInMode(this._projectedCalculationColumns, this._calculationRows);
+      }
+    });
+  }
+
+  private displayViewsInMode(columnsViewContainer: ViewContainerRef, rowsViewContainer: ViewContainerRef) {
```

Comment:
nit: naming is a bit confusing to me. 
suggestions: insertViewsForMode, inserViewsForDisplay, insertViewsForDisplayMode. 

Not a blocker tho.